### PR TITLE
Add LeastSquaresProblem capabilities

### PR DESCRIPTION
**Pull request by jschueller**
- Added LeastSquaresProblem class
- Added NearestPointProblem class (OptimizationProblem(levelValue, levelFunction) is deprecated)
- Added CMinpack class which is an interface to https://github.com/devernay/cminpack, with Levenberge-Marquart algorithm

### CODE REVIEW

File: lib/src/Base/Optim/LeastSquaresProblem.cxx [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+
+  residualFunction_ = residualFunction;
+  const Point center(residualSize);
+  const Point constant(1);
+  const Matrix linear(residualSize, 1);
+  SymmetricTensor quadratic(residualSize, 1);
+  const QuadraticFunction sumQuad(center, constant, linear, quadratic);
+  for (UnsignedInteger i = 0; i < residualSize; ++ i)
+    quadratic(i, i, 0) = 0.5;
+  objective_ = ComposedFunction(sumQuad, residualFunction);
```
Comment (regislebrun) [inline lib/src/Base/Optim/LeastSquaresProblem.cxx]: Not the most efficient way to compute half of the squared norm of a vector, but at least it exists and it works ;-)! I will add a dedicated function to compute u->||u||^2/2, its gradient u->u and its Hessian u->Id later on.

File: lib/src/Base/Optim/NearestPointProblem.cxx [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+  levelFunction_ = levelFunction;
+  dimension_ = levelFunction_.getInputDimension();
+  // Update objective function
+  const Point center(dimension_);
+  const Point constant(1);
+  const Matrix linear(dimension_, 1);
+  const IdentityMatrix identity(dimension_);
+  const SymmetricTensor quadratic(dimension_, 1, *(identity.getImplementation().get()));
+  objective_ = QuadraticFunction(center, constant, linear, quadratic);
```
Comment (regislebrun) [inline lib/src/Base/Optim/NearestPointProblem.cxx]: Same remark as in NonLinearLeastSquaresProblem class

File: python/src/LeastSquaresProblem_doc.i.in [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+.. math::
+
+    \min_{x\in B} f(x) \\
+    g(x) = 0 \\
+    h(x) \ge 0
+
+where *B* is problem's bounds, *f* is the objective function, *g* are equality constraints, and *h* are inequality constraints.
+
+The second form represents a nearest point problem:
```
Comment (regislebrun) [inline python/src/LeastSquaresProblem_doc.i.in]: It is rather a least-squares problem: \min ||f(x)||^2/2

File: python/src/NearestPointProblem_doc.i.in [hunk 4]
```diff
@@ -0,0 +1,68 @@
+%feature("docstring") OT::NearestPointProblem
+"Nearest point problem.
+
+This represents a nearest point problem:
+
+.. math::
+
+    \min_{x} ||x|| \\
```
Comment (regislebrun) [inline python/src/NearestPointProblem_doc.i.in]: Rather write \min ||x||^2/2

File: python/src/LeastSquaresProblem_doc.i.in [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+.. math::
+
+    \min_{x} ||x|| \\
+    g(x) = v
+
+Examples
+--------
+>>> import openturns as ot
+>>> residualFunction = ot.SymbolicFunction(['x1', 'x2'], ['(1-x1)^2+100*(x2-x1^2)^2'])
```
Comment (regislebrun) [inline python/src/LeastSquaresProblem_doc.i.in]: Maybe a function with output dimension > 1?

**Comment by jschueller** [comment]
@regislebrun I added a basic wrapper to cminpack in this branch with Levenberg–Marquardt so you can try it for your work on calibration

**Comment by regislebrun** [comment]
@jschueller Good idea to postpone ROL integration, as it looks a little bit tricky to compile/install out of the box due to its large dependence set and poor internal dependence management...

File: lib/src/Base/Optim/CMinpack.cxx [hunk 6]
```diff
@@ -0,0 +1,479 @@
...
+  if (problem.hasEqualityConstraint())
+    throw InvalidArgumentException(HERE) << "Error: " << getAlgorithmName() << " does not support equality constraints";
+
+#else
+  throw NotYetImplementedException(HERE) << "No CMinpack support";
+#endif
+}
+
+
+int CMinpack::ComputeObjective(void *p, int m, int n, const Scalar *x, Scalar *fvec, Scalar *fjac, int ldfjac, int iflag)
```
Comment (regislebrun) [inline lib/src/Base/Optim/CMinpack.cxx]: m and ldfjac are unused

File: lib/src/Base/Optim/CMinpack.cxx [hunk 7]
```diff
@@ -0,0 +1,479 @@
...
+  if (!algorithm)
+    throw InternalException(HERE) << "p is null";
+  const UnsignedInteger dimension = algorithm->getProblem().getDimension();
+  Point inP(dimension);
+  std::copy(x, x + n, inP.begin());
+
+  if (iflag == 1)
+  {
+    // evaluation
+    Point outP(algorithm->getProblem().getResidualFunction()(inP));
```
Comment (regislebrun) [inline lib/src/Base/Optim/CMinpack.cxx]: const Point

File: lib/src/Base/Optim/CMinpack.cxx [hunk 8]
```diff
@@ -0,0 +1,479 @@
...
+    // track input/outputs
+    algorithm->evaluationInputHistory_.add(inP);
+    algorithm->evaluationOutputHistory_.add(Point(1, 0.5 * outP.normSquare()));
+
+    std::copy(outP.begin(), outP.end(), fvec);
+  }
+  else if (iflag == 2)
+  {
+    // gradient
+    Matrix gradient(algorithm->getProblem().getResidualFunction().gradient(inP).transpose());
```
Comment (regislebrun) [inline lib/src/Base/Optim/CMinpack.cxx]: lines 120 to 128 can be replaced by:
```
    const Point gradient(*algorithm->getProblem().getResidualFunction().gradient(inP).transpose().getImplementation());
    const Scalar norm = gradient.norm();
    LOGINFO(OSS() << "Gradient norm: " << norm);
    std::copy(gradient.begin(), gradient.end(), fjac);
```

**Comment by jschueller** [comment]
@regislebrun only lmder is interesting for us so I removed the algo name parameter
I added box bound constraints support using the tanh variable change proposed in cminpack example.
I guess I'll merge it soon then.

**Comment by regislebrun** [comment]
@jschueller I agree with you, only the nonlinear least-squares algorithm with user-provided gradient is of interest for us.
At one point we were interested into having a multivariate nonlinear equation solver. CMINPACK provides such an algorithm, but of course it is not an optimization algorithm.
IMO it is ready to merge.

**Pull request merged**